livekit-agents/livekit/agents/transcription/tts_forwarder.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import contextlib
+import time
+import uuid
+from collections import deque
+from typing import Callable
+
+from attrs import define
+from livekit import rtc
+
+from .. import tokenize
+from ..log import logger
+
+
+@define
+class TTSOptions:
+    room: rtc.Room
+    participant_identity: str
+    track_id: str
+    language: str
+    speed: float
+    word_tokenizer: tokenize.WordTokenizer
+    sentence_tokenizer: tokenize.SentenceTokenizer
+    hyphenate_word: Callable[[str], list[str]]
+
+
+@define
+class _SegmentData:
+    sentence_stream: tokenize.SentenceStream
+    text: str
+    audio_duration: float
+    avg_speed: float | None
+    processed_hyphenes: int
+
+
+@define
+class _PendingSegments:
+    cur_audio: _SegmentData
+    cur_text: _SegmentData
+    q: deque[_SegmentData]
+
+
+def _uuid() -> str:
+    return str(uuid.uuid4())[:12]
+
+
+class TTSSegmentsForwarder:
+    def __init__(
+        self,
+        opts: TTSOptions,
+    ):
+        self._opts = opts
+        self._main_task = asyncio.create_task(self._run())
+
+        # current segment where the user may still be pushing text & audio
+        first_segment = self._create_segment()
+        segments_q = deque()
+        segments_q.append(first_segment)
+
+        self._pending_segment = _PendingSegments(
+            cur_audio=first_segment,
+            cur_text=first_segment,
+            q=segments_q,
+        )
+
+        self._seg_queue = asyncio.Queue[_SegmentData | None]()
+        self._seg_queue.put_nowait(first_segment)
+
+        # the forwarding of the transcription is started as soon as the user push the first audio frame.
+        self._start_future = asyncio.Future()
+
+    def validate(self) -> None:
+        """Validate must be called once we started to playout the audio to the user"""
+        with contextlib.suppress(asyncio.InvalidStateError):
+            self._start_future.set_result(None)
+
+    def push_audio(self, frame: rtc.AudioFrame | None, **kwargs) -> None:
+        if frame is not None:
+            frame_duration = frame.samples_per_channel / frame.sample_rate
+            cur_seg = self._pending_segment.cur_audio
+            cur_seg.audio_duration += frame_duration
+        else:
+            self.mark_audio_segment_end()
+
+    def mark_audio_segment_end(self) -> None:
+        try:
+            # get last ended segment (text always end before audio)
+            seg = self._pending_segment.q.popleft()
+        except IndexError:
+            raise IndexError(
+                "mark_audio_segment_end called before any mark_segment_end"
+            )
+
+        seg.avg_speed = len(self._calc_hyphenes(seg.text)) / seg.audio_duration
+        self._pending_segment.cur_audio = self._pending_segment.q[0]
+
+    def push_text(self, text: str | None) -> None:
```

Comment:
can we add docstrings on usage? also I think for non-streaming LLMs, users would need to explicitly call `end_text_segment`.. this feels cleaner to call `push_text(text, end_segment=True)`, versus assuming None ends a segment..

Reply:
TTS has the same method, I think it was OK to do it this way because we have mark_end_segment